template/template.go

```diff
@@ -157,34 +181,37 @@ func (e *executer) resolveComplexMessage(message ast.ComplexMessage) error {
 	return resolutionErr
 }
 
-func (e *executer) resolveDeclarations(declarations []ast.Declaration) error {
-	m := make(map[ast.Value]struct{}, len(declarations))
+func newLiteral(v any) ast.Value { //nolint:ireturn
+	switch t := v.(type) {
+	case string:
+		return ast.QuotedLiteral(t)
+	case float64:
+		return ast.NumberLiteral(t)
+	case int:
+		return ast.NumberLiteral(t)
+	}
+
+	return nil
+}
 
+func (e *executer) resolveDeclarations(declarations []ast.Declaration) error {
 	for _, decl := range declarations {
 		switch d := decl.(type) {
 		case ast.ReservedStatement:
 			return fmt.Errorf("%w", mf2.ErrUnsupportedStatement)
 		case ast.LocalDeclaration:
-			m[d.Variable] = struct{}{}
-
-			resolved, err := e.resolveExpression(d.Expression)
-			// if can't resolve the expression, leave it as unresolved, e.g. {$foo}
-			e.variables[string(d.Variable)] = resolved
-
-			if err != nil {
-				return fmt.Errorf("local declaration: %w", err)
-			}
-
+			e.variables[string(d.Variable)] = newVariable(d.Expression, nil)
 		case ast.InputDeclaration:
-			m[d.Operand] = struct{}{}
+			name := string(d.Operand.(ast.Variable)) //nolint: forcetypeassert // always ast.Variable
 
-			resolved, err := e.resolveExpression(ast.Expression(d))
-			// if can't resolve the expression, leave it as unresolved, e.g. {$foo}
-			e.variables[string(d.Operand.(ast.Variable))] = resolved //nolint: forcetypeassert // Will always be a variable.
-
-			if err != nil {
-				return fmt.Errorf("input declaration: %w", err)
+			val, ok := e.variables[name]
+			if !ok {
+				return mf2.ErrUnresolvedVariable
 			}
+
+			expr := ast.Expression(d)
+			expr.Operand = newLiteral(val)
```

Comment:
If input contains the `time.Location` type, what happens? How do we handle this scenario?

Reply:
We cannot store this value in any literal in the `parse` package.

Reply:
I guess that will cause some mild panic in some part of the template execution. In all other cases input is either string or number (float64 or int).

Reply:
We require to support `time.Location` and others -  without serialization.

Reply:
Values from input eventually get used either as Operand for an Annotation or value for Option - in neither case `time.Location` is valid type. It most likely gets passed as a string (QuotedLiteral) and only the function (Annotation) interprets it as `time.Location`.

Reply:
What MF2 expects in this case? is $tz input?

```go
tpl.Parse(`{{ |2024-08-01| :date timeZone={$tz} }}`)
tpl.Sprint(map[string]any{"tz": location.UTC})
```

Reply:
Yes, `tz` is input. In `resolveValue()` it's going to grab the value from variables and pass it into registry function Format() which will interpret it as needed. Looks like in this scenario the location type doesn't get serialized into string. I'll double check the MF2 spec docs.

Reply:
#190 will probably solve it